community/kernel/src/main/java/org/neo4j/kernel/impl/index/labelscan/LabelScanValueIterator.java

```diff
@@ -0,0 +1,127 @@
+ * iterate over each set bit, returning actual node ids, i.e. {@code nodeIdRange+bitOffset}.
+ *
+ * The provided {@link RawCursor} is managed externally, e.g. {@link NativeLabelScanReader},
+ * this because implemented interface lacks close-method.
+ */
+class LabelScanValueIterator extends PrimitiveLongCollections.PrimitiveLongBaseIterator
+{
+    /**
+     * Size of nodeId ranges, i.e. how many node ids one {@link LabelScanValue} can hold at most.
+     */
+    private final int rangeSize;
+
+    /**
+     * {@link RawCursor} to lazily read new {@link LabelScanValue} from.
+     */
+    private final RawCursor<Hit<LabelScanKey,LabelScanValue>,IOException> cursor;
+
+    /**
+     * Current base nodeId, i.e. the {@link LabelScanKey#idRange} of the current {@link LabelScanKey}.
+     */
+    private long baseNodeId;
+
+    /**
+     * Bit set of the current {@link LabelScanValue}.
+     */
+    private long bits;
+
+    /**
+     * LabelId of previously retrieved {@link LabelScanKey}, for debugging and asserting purposes.
+     */
+    private int prevLabel = -1;
+
+    /**
+     * IdRange of previously retrieved {@link LabelScanKey}, for debugging and asserting purposes.
+     */
+    private long prevRange = -1;
+
+    LabelScanValueIterator( int rangeSize, RawCursor<Hit<LabelScanKey,LabelScanValue>,IOException> cursor )
+    {
+        this.rangeSize = rangeSize;
+        this.cursor = cursor;
+    }
+
+    /**
+     * @return next node id in the current {@link LabelScanValue} or, if current value exhausted,
+     * goes to next {@link LabelScanValue} from {@link RawCursor}. Returns {@code true} if next node id
+     * was found, otherwise {@code false}.
+     */
+    @Override
+    protected boolean fetchNext()
+    {
+        while ( true )
+        {
+            if ( bits != 0 )
+            {
+                return nextFromCurrent();
+            }
+
+            try
+            {
+                if ( !cursor.next() )
+                {
+                    return false;
+                }
+            }
+            catch ( IOException e )
+            {
+                throw new RuntimeException( e );
+            }
+
+            Hit<LabelScanKey,LabelScanValue> hit = cursor.get();
+            baseNodeId = hit.key().idRange * rangeSize;
+            bits = hit.value().bits;
+
+            assert keysInOrder( hit.key() );
+        }
+    }
+
+    private boolean keysInOrder( LabelScanKey key )
+    {
+        assert key.labelId >= prevLabel : "Expected to get ordered results, got " + key +
+                " where previous label was " + prevLabel;
+        assert key.idRange > prevRange : "Expected to get ordered results, got " + key +
+                " where previous range was " + prevRange;
+        prevLabel = key.labelId;
+        prevRange = key.idRange;
+        // Made as a method returning boolean so that it can participate in an assert call.
```

Comment:
inside `fetchNext` this method is used as an assertion, it's simply a helpful fail-fast debug thing we've added.